contracts/utils/Nonces.sol

```diff
@@ -1,22 +1,74 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.19;
 
+import {Context} from "./Context.sol";
+
 /**
  * @dev Provides tracking nonces for addresses. Nonces will only increment.
  */
-abstract contract Nonces {
+abstract contract Nonces is Context {
+    bytes32 internal constant DEFAULT_TYPEHASH = 0;
+
     /**
      * @dev The nonce used for an `account` is not the expected current nonce.
      */
     error InvalidAccountNonce(address account, uint256 currentNonce);
 
-    mapping(address => uint256) private _nonces;
+    error InvalidFastForward(address account, uint256 currentNonce);
+
+    // signer → typehash → beneficiary (track) → nonce
+    mapping(address => mapping(bytes32 => mapping(bytes32 => uint256))) private _nonces;
 
     /**
-     * @dev Returns an the next unused nonce for an address.
+     * @dev Returns the next unused nonce for an address.
      */
     function nonces(address owner) public view virtual returns (uint256) {
-        return _nonces[owner];
+        return operationIds(DEFAULT_TYPEHASH, owner, 0);
+    }
+
+    /**
+     * @dev Returns the next unused nonce for an address and a typehash.
+     */
+    function operationNonces(bytes32 typehash, address signer) public view virtual returns (uint256) {
+        return operationIds(typehash, signer, 0);
+    }
+
+    /**
+     * @dev Returns the next unused nonce for an address, a typehash and a beneficiary.
+     */
+    function operationIds(bytes32 typehash, address signer, bytes32 beneficiary) public view virtual returns (uint256) {
```

Comment:
"Beneficiary" is not the right term, we should probably use "track"?

Reply:
Again, name comes from 6077. I like "track"